### PR TITLE
SBERDOMA-192 classifier for tickets

### PR DESCRIPTION
**Pull request by dkoviazin**
Dependant lvls

image

View page

image

TODO
1. Server validation

### CODE REVIEW

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx [hunk 1]
```diff
@@ -7,7 +7,8 @@ import React, { useEffect, useRef, useState } from 'react'
 import { ITicketFormState } from '@condo/domains/ticket/utils/clientSchema/Ticket'
 import { FormWithAction } from '@condo/domains/common/components/containers/FormList'
 import { PropertyAddressSearchInput } from '@condo/domains/property/components/PropertyAddressSearchInput'
-import { searchEmployee, searchTicketClassifier } from '../../utils/clientSchema/search'
+import { searchEmployee } from '../../utils/clientSchema/search'
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/BaseTicketForm/index.tsx:10]: @domains/...

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+        // We need to remove search classifiers when rules start to work
+        setSearchClassifiers([])
+    }
+
+    useEffect(() => {
+        optionsRef.current = uniqBy([...classifiers, ...searchClassifiers], 'id')
+    }, [classifiers, searchClassifiers])
+
+    const SelectComponent = useMemo(() => {
+        const SelectComponentWrapper: ClassifierSelectComponent = (props) => {
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:67]: move outside hook scope

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                        optionsRef.current.map(classifier => (
+                            <Option value={classifier.id} key={classifier.id} title={classifier.name}>{classifier.name}</Option>
+                        ))
+                    }
+
+                </Select>
+            )
+        }
+        return SelectComponentWrapper
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:94]: remove eslint disable, add missed deps or rewrite hook logic

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    categoryClassifier,
+    descriptionClassifier,
+} }: ITicketThreeLevelsClassifierHookInput): ITicketThreeLevelsClassifierHookOutput => {
+    const intl = useIntl()
+    const PlaceClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.PlaceLabel' })
+    const CategoryClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.CategoryLabel' })
+    const DescriptionClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.DescriptionLabel' })
+    const threeLvlSelectState = useRef({ id: classifierRule, place: null, category:null, description: null })
+    const client = useApolloClient()
+    const helper = new ClassifiersQueryLocal(client)
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:120]: move to useMemo, or outside hook scope

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+    })
+
+    const Setter = {
+        place: placeSet,
+        category: categorySet,
+        description: descriptionSet,
+    }
+
+    useEffect(() => {
+        helper.init().then(_ => {
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:169]: add readable name to 'helper' variable

Comment (Dimitreee) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:169]:
```suggestion
        helper.init().then(_ => {
```

What is ```_ => {}``` style?

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        if (!state.place && state.category) {
+            openSelect(placeRef)
+        } else if (!state.category && state.place) {
+            openSelect(categoryRef)
+        }
+    }
+
+
+    const ClassifiersEditorComponent = useMemo(() => {
+        const ClassifiersEditorWrapper: React.FC<{ form, disabled }> = ({ form, disabled }) => {
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:288]: move outside hook scope

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+                helper.search('', 'category').then(categories => {
+                    Setter.category.all(categories)
+                })
+            }
+        })
+        return () => {
+            // clear all loaded data from helper
+            helper.clear()
+        }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:189]: remove eslint disable or rewrite hook logic

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    }, [])
+
+    const load3levels = async () => {
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:192]:
```suggestion
    const loadLevels = async () => {
```

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+        const { place, category, description } = threeLvlSelectState.current
+        console.log('load3levels: >>> threeLvlSelectState.current', threeLvlSelectState.current)
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:194]: remove log

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+        const loadedRules = await Promise.all([
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:195]: Promise.alll?

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+        ].map(selector => {
+            const { type, ...querySelectors } = selector
+            return new Promise<[string, Options[]]>(resolve => {
+                const query = {}
+                for (const key in querySelectors) {
+                    if (querySelectors[key]) {
+                        query[key] = { id: querySelectors[key] }
+                    }
+                }
+                console.log('load3levels - query: ', query)
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:208]: remove log

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+            })
+        }))
+        const result = Object.fromEntries(loadedRules)
+        return result
+    }
+
+    const openSelect = (ref) => {
+        if (ref.current) {
+            ref.current.blur()
+            setTimeout(ref.current.focus, 0)
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:219]: Why we couldn't use default aunt Select API for this?

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+        const matchingRules = await helper.findRules(query)
+        if (matchingRules.length === 1) {
+            threeLvlSelectState.current = { ...threeLvlSelectState.current, id: matchingRules[0].id }
+        } else if (threeLvlSelectState.current.place && threeLvlSelectState.current.category) {
+            const withEmptyDescription = find(matchingRules, { description: null })
+            if (withEmptyDescription){
+                threeLvlSelectState.current = { ...threeLvlSelectState.current, id: withEmptyDescription.id }
+            }
+        }
+        console.log('SET FIELDS: ', [
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:240]: remove log

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+        ])
+    }
+
+    const updateLevels = async (selected = {}, maxUpdates = 2 ) => {
+        threeLvlSelectState.current = { ...threeLvlSelectState.current, ...selected }
+        const options = await load3levels()
+        const state = threeLvlSelectState.current
+        const updateEmptyState = {}
+        Object.keys(Setter).forEach(type => {
+            const isExisted = options[type].find(x => x.id === state[type])
```
Comment (Dimitreee) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:260]:
```suggestion
            const isExisted = options[type].find(option => option.id === state[type])
```

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx (hunk 1 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/BaseTicketForm/index.tsx:10]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 8 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:192]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 9 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:194]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 10 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:195]: yes, we want 3 queries to run simultaneously

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 11 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:208]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 13 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:240]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 14 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:260]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 3 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx]: This hook needs to be run only once on create - so it has empty deps

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 5 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx]: Fixed

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 7 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:188]: This hook needs to be run only once on create - so it has empty deps

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 12 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:216]: we need to set the focus for text typing and open select in some cases

Default Ant'd API allows only to change the way select opens -> on focus = I use this to open it on manual focus call
and to set the default open - Initial open state of dropdown - we don't need to open it on init
the open property does not set a focus for typing

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 15]
```diff
@@ -0,0 +1,310 @@
+import React, { useEffect, useState, useRef, useMemo } from 'react'
+import { Select, Input, Col, Form } from 'antd'
+const { Option } = Select
+import { useApolloClient } from '@core/next/apollo'
+import { useIntl } from '@core/next/intl'
+
+
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:7]:
```suggestion
```

**Review by AntonAL — COMMENTED** [review]
Should not 2-nd and 3-rd classifier select inputs cleared, when I clear first select?

**Review by AntonAL — COMMENTED** [review]
`DELETE` works without extra conditions, SQL in `down` method can be simplified in this way:

```
DELETE FROM public."TicketPlaceClassifier";
DELETE FROM public."TicketCategoryClassifier";
DELETE FROM public."TicketDescriptionClassifier";
DELETE FROM public."TicketClassifierRule";
```

PS: GitHub have a bug with displaying of my suggested changes in migration file ;)


Fixed

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 15 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx]: Done

**Review by sitozzz — APPROVED** [review]

File: apps/condo/domains/ticket/access/TicketDescriptionClassifier.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+const { throwAuthenticationError } = require('@condo/domains/common/utils/apolloErrorFormatter')
+
+async function canReadTicketDescriptionClassifiers ({ authentication: { item: user } }) {
+    if (!user) return throwAuthenticationError()
+    if (user.isAdmin) return {}
+    return {}
+}
+
+async function canManageTicketDescriptionClassifiers ({ authentication: { item: user }, originalInput, operation, itemId }) {
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketDescriptionClassifier.js:13]: canManageTicketProblemClassifiers

File: apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx [hunk 17]
```diff
@@ -159,6 +159,30 @@ const useChangedFieldMessagesOf = (ticketChange) => {
...
+        {
+            title: 'categoryClassifierDisplayName',
+            messages: {
+                add: ClassifierAddMessage,
+                change: ClassifierChangeMessage,
+                remove: ClassifierRemoveMessage,
+            },
+        },
+        {
+            title: 'descriptionClassifierDisplayName',
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx:179]: problem

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+    classifierRule,
+    placeClassifier,
+    categoryClassifier,
+    descriptionClassifier,
+} }: ITicketThreeLevelsClassifierHookInput): ITicketThreeLevelsClassifierHookOutput => {
+    const intl = useIntl()
+    const PlaceClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.PlaceLabel' })
+    const CategoryClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.CategoryLabel' })
+    const DescriptionClassifierLabel = intl.formatMessage({ id: 'component.ticketclassifier.DescriptionLabel' })
+    const ruleRef = useRef({ id: classifierRule, place: null, category:null, description: null })
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:120]:
```suggestion

    const ruleRef = useRef({ id: classifierRule, place: null, category:null, description: null })
```

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+        return result
+    }
+
+    const openSelect = (ref) => {
+        if (ref.current) {
+            ref.current.blur()
+            setTimeout(ref.current.focus, 0)
+        }
+    }
+    // Every time user choose some option from select we are trying to find what exact rule is matching for this combination
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:235]:
```suggestion

    // Every time user choose some option from select we are trying to find what exact rule is matching for this combination
```

File: apps/condo/domains/ticket/gql.js [hunk 20]
```diff
@@ -97,6 +92,18 @@ const TICKET_CHANGE_DATA_FIELDS = [
...
     'classifierDisplayNameTo',
+    'placeClassifierIdFrom',
+    'placeClassifierIdTo',
+    'placeClassifierDisplayNameFrom',
+    'placeClassifierDisplayNameTo',
+    'categoryClassifierIdFrom',
+    'categoryClassifierIdTo',
+    'categoryClassifierDisplayNameFrom',
+    'categoryClassifierDisplayNameTo',
+    'descriptionClassifierIdFrom',
```
Comment (pahaz) [inline apps/condo/domains/ticket/gql.js:103]: problem

File: apps/condo/domains/ticket/gql.js [hunk 21]
```diff
@@ -134,19 +137,29 @@ const TICKET_ANALYTICS_REPORT_MUTATION = gql`
...
 const ResidentTicket = generateGqlQueries('ResidentTicket', RESIDENT_TICKET_FIELDS)
 
+const TICKET_PLACE_CLASSIFIER_FIELDS = `{ organization { id } name ${COMMON_FIELDS} }`
+const TicketPlaceClassifier = generateGqlQueries('TicketPlaceClassifier', TICKET_PLACE_CLASSIFIER_FIELDS)
+
+const TICKET_CATEGORY_CLASSIFIER_FIELDS = `{ organization { id } name ${COMMON_FIELDS} }`
+const TicketCategoryClassifier = generateGqlQueries('TicketCategoryClassifier', TICKET_CATEGORY_CLASSIFIER_FIELDS)
+
+const TICKET_DESCRIPTION_CLASSIFIER_FIELDS = `{ organization { id } name ${COMMON_FIELDS} }`
+const TicketDescriptionClassifier = generateGqlQueries('TicketDescriptionClassifier', TICKET_DESCRIPTION_CLASSIFIER_FIELDS)
```
Comment (pahaz) [inline apps/condo/domains/ticket/gql.js:147]: problem

File: apps/condo/domains/ticket/gql.js [hunk 22]
```diff
@@ -162,6 +175,10 @@ module.exports = {
     TicketComment,
     TICKET_ANALYTICS_REPORT_MUTATION,
     SHARE_TICKET_MUTATION,
+    TicketPlaceClassifier,
+    TicketCategoryClassifier,
+    TicketDescriptionClassifier,
```
Comment (pahaz) [inline apps/condo/domains/ticket/gql.js:180]: problem

File: apps/condo/domains/ticket/access/TicketDescriptionClassifier.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Generated by `createschema ticket.TicketDescriptionClassifier 'organization?:Relationship:Organization:CASCADE;name:Text;'`
+ */
+
+const { throwAuthenticationError } = require('@condo/domains/common/utils/apolloErrorFormatter')
+
+async function canReadTicketDescriptionClassifiers ({ authentication: { item: user } }) {
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketDescriptionClassifier.js:7]: problem

File: apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx (hunk 17 above)
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx:179]: problem

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 18 above)
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:120]:
```suggestion

    const ruleRef = useRef({ id: classifierRule, place: null, category:null, description: null })
```

File: apps/condo/domains/ticket/gql.js [hunk 24]
```diff
@@ -3,33 +3,28 @@
  * In most cases you should not change it by hands
  * Please, don't remove `AUTOGENERATE MARKER`s
  */
-
 const { generateGqlQueries } = require('@condo/domains/common/utils/codegeneration/generate.gql')
 const { gql } = require('graphql-tag')
 const COMMON_FIELDS = 'id dv sender v deletedAt newId createdBy { id name } updatedBy { id name } createdAt updatedAt'
 
-const TICKET_FIELDS = `{ organization { id name } property { id name address } unitName sectionName floorName status { id name type organization { id } colors } statusReopenedCounter statusUpdatedAt statusReason number client { id name } clientName clientEmail clientPhone contact { id name } operator { id name } assignee { id name } executor { id name } watchers { id name } classifier { id name organization { id } parent { id name } } details related { id details } isEmergency isPaid meta source { id name type } sourceMeta ${COMMON_FIELDS} }`
+const THREE_LVL_CLASSIFIER_FIELDS = 'placeClassifier { id name } categoryClassifier { id name } descriptionClassifier { id name } classifierRule { id }'
```
Comment (pahaz) [inline apps/condo/domains/ticket/gql.js:10]: you forgot to add `{` and `}`

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 25]
```diff
@@ -147,20 +147,55 @@ const Ticket = new GQLListSchema('Ticket', {
...
             type: Relationship,
             ref: 'TicketClassifier',
             // Classifier can be null because mobile app users don't pass it when creating a Ticket
-            kmigratorOptions: { null: true, on_delete: 'models.SET_NULL' },
+            isRequired: false,
+            knexOptions: { isNotNullable: false }, // Required relationship only!
+            kmigratorOptions: { null: true, on_delete: 'models.PROTECT' },
+        },
+        // TODO(zuch): make it required
+        placeClassifier: {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:160]: just `place`

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 26]
```diff
@@ -147,20 +147,55 @@ const Ticket = new GQLListSchema('Ticket', {
...
+            schemaDoc: 'Describe where incident took place',
+            type: Relationship,
+            ref: 'TicketPlaceClassifier',
+            isRequired: false,
+            knexOptions: { isNotNullable: false },
+            kmigratorOptions: { null: true, on_delete: 'models.PROTECT' },
+        },
+        // TODO(zuch): make it required
+        categoryClassifier: {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:169]: `category`

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 27]
```diff
@@ -147,20 +147,55 @@ const Ticket = new GQLListSchema('Ticket', {
...
+            schemaDoc: 'Describe type of work needed',
+            type: Relationship,
+            ref: 'TicketCategoryClassifier',
+            isRequired: false,
+            knexOptions: { isNotNullable: false },
+            kmigratorOptions: { null: true, on_delete: 'models.PROTECT' },
+        },
+        problemClassifier: {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:177]: problem

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 28]
```diff
@@ -147,20 +147,55 @@ const Ticket = new GQLListSchema('Ticket', {
...
+            knexOptions: { isNotNullable: false },
+            kmigratorOptions: { null: true, on_delete: 'models.PROTECT' },
         },
-
         // description / title
         details: {
             schemaDoc: 'Text description of the issue. Maybe written by a user or an operator',
             type: Text,
-            isRequired: true,
+            isRequired: false,
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:198]: we should have required text

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 29]
```diff
@@ -257,7 +291,8 @@ const Ticket = new GQLListSchema('Ticket', {
             return resolvedData
         },
         validateInput: ({ resolvedData, existingItem, addValidationError, operation }) => {
-            if (!hasRequestAndDbFields(['dv', 'sender'], ['organization', 'source', 'status', 'details'], resolvedData, existingItem, addValidationError)) return
+            // Todo(zuch): add placeClassifier, categoryClassifier and classifierRule
+            if (!hasRequestAndDbFields(['dv', 'sender'], ['organization', 'source', 'status'], resolvedData, existingItem, addValidationError)) return
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:295]: add details

File: apps/condo/domains/ticket/schema/TicketCategoryClassifier.js [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+
+const { Text } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
+const access = require('@condo/domains/ticket/access/TicketCategoryClassifier')
+const { COMMON_AND_ORGANIZATION_OWNED_FIELD } = require('../../../schema/_common')
+
+const TicketCategoryClassifier = new GQLListSchema('TicketCategoryClassifier', {
+    schemaDoc: 'Describes what type of work needs to be done to fix incident ',
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketCategoryClassifier.js:13]:
```suggestion
    schemaDoc: 'Describes what type of work needs to be done to fix incident',
```

File: apps/condo/domains/ticket/schema/Ticket.js [hunk 31]
```diff
@@ -147,20 +147,55 @@ const Ticket = new GQLListSchema('Ticket', {
...
+        problemClassifier: {
+            schemaDoc: 'Details of incident',
+            type: Relationship,
+            ref: 'TicketProblemClassifier',
+            isRequired: false,
+            knexOptions: { isNotNullable: false },
+            kmigratorOptions: { null: true, on_delete: 'models.PROTECT' },
+        },
+        // TODO(zuch): make it required
+        classifierRule: {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/Ticket.js:186]: Maybe it's ok to rename it to classifier after the migration? (to think)

File: apps/condo/domains/ticket/schema/TicketClassifierRule.js [hunk 32]
```diff
@@ -0,0 +1,50 @@
...
+
+const { Relationship } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
+const access = require('@condo/domains/ticket/access/TicketClassifierRule')
+const { COMMON_AND_ORGANIZATION_OWNED_FIELD } = require('../../../schema/_common')
+
+const TicketClassifierRule = new GQLListSchema('TicketClassifierRule', {
+    schemaDoc: 'Rules for  all possible valid combinations of classifiers',
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketClassifierRule.js:13]:
```suggestion
    schemaDoc: 'Rules for all possible valid combinations of classifiers',
```

File: apps/condo/domains/ticket/schema/TicketClassifierRule.js [hunk 33]
```diff
@@ -0,0 +1,50 @@
+/**
+ * Generated by `createschema ticket.TicketClassifierRule 'organization?:Relationship:Organization:CASCADE;place?:Relationship:TicketPlaceClassifier:PROTECT;category?:Relationship:TicketCategoryClassifier:PROTECT;problem?:Relationship:TicketProblemClassifier:PROTECT;'`
+ */
+
+const { Relationship } = require('@keystonejs/fields')
+const { GQLListSchema } = require('@core/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@core/keystone/plugins')
+const { SENDER_FIELD, DV_FIELD } = require('@condo/domains/common/schema/fields')
+const access = require('@condo/domains/ticket/access/TicketClassifierRule')
+const { COMMON_AND_ORGANIZATION_OWNED_FIELD } = require('../../../schema/_common')
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketClassifierRule.js:10]: Wow! we need to move it to `domain/organization/schema/fields.js` (like common fields)

File: apps/condo/domains/ticket/utils/clientSchema/Ticket.ts [hunk 34]
```diff
@@ -10,9 +10,9 @@ import { generateReactHooks } from '@condo/domains/common/utils/codegeneration/g
...
-const RELATIONS = ['status', 'client', 'contact', 'operator', 'assignee', 'classifier', 'organization', 'source', 'property', 'executor', 'related']
-
+const FIELDS = ['id', 'deletedAt', 'createdAt', 'updatedAt', 'createdBy', 'updatedBy', 'organization', 'statusReopenedCounter', 'statusReason', 'statusUpdatedAt', 'status', 'number', 'client', 'clientName', 'clientEmail', 'clientPhone', 'contact', 'unitName', 'sectionName', 'floorName', 'watchers', 'operator', 'assignee', 'classifier', 'placeClassifier', 'categoryClassifier', 'descriptionClassifier', 'classifierRule', 'details', 'related', 'isEmergency', 'isPaid', 'meta', 'source', 'property', 'executor']
+const RELATIONS = ['status', 'client', 'contact', 'operator', 'assignee', 'classifier', 'organization', 'source', 'property', 'executor', 'related', 'placeClassifier', 'categoryClassifier', 'descriptionClassifier', 'classifierRule']
+const DISCONNECT_ON_NULL = ['descriptionClassifier']
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/clientSchema/Ticket.ts]: problem?

File: apps/condo/domains/ticket/access/TicketDescriptionClassifier.js (hunk 16 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/access/TicketDescriptionClassifier.js]: Done

File: apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx (hunk 17 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketChanges/TicketChange.tsx]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 18 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx]: ruleRef - problem = fixed

File: apps/condo/domains/ticket/gql.js (hunk 24 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/gql.js]: They belong to ticket

File: apps/condo/domains/ticket/schema/Ticket.js (hunk 28 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/schema/Ticket.js]: done

File: apps/condo/domains/ticket/schema/Ticket.js (hunk 29 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/schema/Ticket.js]: done

File: apps/condo/domains/ticket/schema/TicketCategoryClassifier.js (hunk 30 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/schema/TicketCategoryClassifier.js]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 3 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx]: Done

File: apps/condo/domains/ticket/components/TicketClassifierSelect.tsx (hunk 19 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/components/TicketClassifierSelect.tsx:235]: Done

File: apps/condo/domains/ticket/schema/TicketClassifierRule.js (hunk 32 above)
Comment (dkoviazin) [reply apps/condo/domains/ticket/schema/TicketClassifierRule.js:13]: Done

**Pull request merged**